File: basic types/null/null.go
// 空值
// go语言有空值的概念，引用类型的默认值为nil.
// 引用类型包括interface, map, channel等

package main

var array = [...]int{1, 2, 3, 4, 5}
var index = [...]int{0, 2, 4, 6}

func get_element_by_index(ind int) map[int]int {
	if ind < 0 || ind >= len(array) {
		return nil
	}
	return map[int]int{array[ind]: array[ind]}
}

func main() {
	sum := 0
	for _, value := range index {
		cur := get_element_by_index(value)
		if cur != nil { // 需要检查函数返回的变量是否为空，不检查也可以编译，但程序有崩溃的危险
			for key := range cur {
				sum += key
			}
		}
	}
	println("sum: ", sum)
}

File: basic types/null/null.rs
// 空值
// rust语言没有空值的概念,只有可空类型。
// Some(3) => 存在值，值为3
// None  => 不存在值
static ARRAY: [isize;5] = [1, 2, 3, 4, 5];
static INDEX: [isize;4] = [0, 2, 4, 6];

fn get_element_by_index(ind: isize) -> Option<isize> {
    if ind < 0 || ind as usize >= ARRAY.len() {
        return None;
    }
    return Some(ARRAY[ind as usize]);
}

fn main() {
    let mut sum = 0;
    for &value in INDEX.iter() {
        let cur = get_element_by_index(value);
        if let Some(v) = cur { // 需要显式检查函数返回的值，否则因为类型不同无法编译
            sum += v;
        }
    }
    println!("sum: {}", sum);
}
